athena/etc/gettime/gettime.8
.\" $Id: gettime.8,v 1.1 1997-12-15 00:35:49 cfields Exp $
.\"
.\" Copyright 1997 by the Massachusetts Institute of Technology.
.\"
.\" Permission to use, copy, modify, and distribute this
.\" software and its documentation for any purpose and without
.\" fee is hereby granted, provided that the above copyright
.\" notice appear in all copies and that both that copyright
.\" notice and this permission notice appear in supporting
.\" documentation, and that the name of M.I.T. not be used in
.\" advertising or publicity pertaining to distribution of the
.\" software without specific, written prior permission.
.\" M.I.T. makes no representations about the suitability of
.\" this software for any purpose.  It is provided "as is"
.\" without express or implied warranty.
.\"
.TH GETTIME 8 "13 Dec 1997"
.SH NAME
gettime \- synchronize system clock with remote host
.SH SYNOPSIS
.B gettime
[
.B \-g granularity
] [
.B \-s
] 
.B hostname
.fi
.SH DESCRIPTION
.I gettime
is a simple program used to retrieve the time from a remote host
(specified by the required
.B hostname
argument) using the UDP Time Protocol defined in RFC868. It displays
the time acquired on the standard output. If the
.B \-s
option is specified and
.I gettime
is being run as root, the clock of the local workstation will also be
set. If the
.B \-g
option is specified as well, the time will only be set if the
workstation clock is more than
.B granularity
seconds out of synchronization with the remote host.
.I gettime
attempts to retrieve the time five times, with a five second timeout
on each attempt.

.SH AUTHOR
Craig Fields, MIT Information Systems
.br
Copyright (c) 1997 Massachusetts Institute of Technology


doc/news
This file is intended to keep track of user-visible changes to the
release, for the construction of the Athena release notes.

Changes from release 8.0 to release 8.1:

	* Athena version numbers are now dotted triplets, e.g. 8.1.0
	  or 8.1.12.  Paths which include all or part of the Athena
	  version number now take the form "pathname-version",
	  e.g. "srvd-8.1" or "dash.menus-8.1"

	* In release 8.0, running "xmkmf" with the default path would
	  generate a Makefile which looked for its include files and
	  libraries in /mit/x11, which could be one of several
	  lockers.  Now it is preferrable to use the vendor X
	  libraries, so this bit of glue is going away.  There will be
	  no /usr/athena/bin/xmkmf any more, nor any symlinks in
	  /usr/athena pointing to the x11 locker.

	  Users can still do something like "add -f x11r5" and then
	  "xmkmf", and things should work, if the x11r5 locker is
	  built for their platform.

	* Athena bitmaps have moved to /usr/athena/include/X11/bitmaps.
	  Previously, they were in /usr/athena/lib/X11/bitmaps before
	  because of the /mit/x11 glue.

	* The symlink /bin/csh -> /bin/athena/tcsh has gone away.  The
	  update to 8.0 will fix /bin/csh references in /etc/passwd
	  and /etc/passwd.local to refer to /bin/athena/tcsh instead.

	* emacs18 has gone away.

	* AFS includes and libraries are installed in
	  /usr/afsws/include and /usr/afsws/lib.

	* wcl includes and libraries are installed in /usr/athena.

	* The basic krb5 headers and libraries are installed in
	  /usr/athena.

	* emacs has been upgraded from 19.30 to 19.34.  There should
	  be no significant user-visible changes.

	* RCS has been upgraded from version 4 to version 5.7.

	* x3270 has been upgraded from version 1.2 to version 3.1.1.3.

	* sysinfo is now available under Irix.

	* /bin/athena/tcsh now has Hesiod support enabled, which means
	  that "~lockername" will work in a pathname in most cases.
	  (It won't work for users with local passwd information who
	  don't have their lockers attached.)

	* For improved ease of debugging, many binaries are installed
	  with symbol tables, and compiled without optimization.
	  Therefore, the contents of /bin/athena and /usr/athena/bin
	  may be somewhat larger, and the binaries may consume more
	  CPU cycles, although it's unlikely that anyone will notice
	  the speed difference.


athena/etc/gettime/gettime.c
/*
 * gettime
 *
 * Retrieve the time from a remote host and print it in ctime(3)
 * format. The time is acquired via the protocol described in
 * RFC868. The name of the host to be queried is specified on the
 * command line, and if the -s option is also specified, and gettime
 * is run as root, the clock on the local workstation will also be
 * set. gettime makes five attempts at getting the time, each with
 * a five second timeout.
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <time.h>
#include <errno.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <sys/time.h>
#include <netinet/in.h>
#include <netdb.h>

static char rcsid[] = "$Id: gettime.c,v 1.12 1997-12-13 23:07:06 cfields Exp $";

#define UNIX_OFFSET_TO_1900 ((70 * 365UL + 17) * 24 * 60 * 60)
#define TIME_SERVICE "time"
#define MAX_TRIES 5
#define TIMEOUT 5

char *program_name;

void myperror(char *what, int errnum)
{
  fprintf(stderr, "%s: %s: %s\n", program_name, what, strerror(errnum));
  exit(1);
}

int main(int argc, char **argv)
{
  struct servent *service_info;
  struct hostent *host_info = NULL;
  struct sockaddr_in time_address;
  char *time_hostname, buffer[20];
  int tries, time_socket, result;
  fd_set read_set;
  struct timeval timeout, current_time;
  struct timezone current_timezone;
  time_t now;
  int c, setflag = 0, errflag = 0;

  /* Set up our program name. */
  if (argv[0] != NULL)
    {
      program_name = strrchr(argv[0], '/');
      if (program_name != NULL)
	program_name++;
      else
	program_name = argv[0];
    }
  else
    program_name = "gettime";

  /* Parse arguments. */
  while ((c = getopt(argc, argv, "s")) != EOF)
    {
      switch (c)
	{
	case 's':
	  setflag = 1;
	  break;
	case '?':
	  errflag = 1;
	  break;
	}
    }

  if (errflag || optind + 1 != argc)
    {
      fprintf(stderr, "usage: %s [-s] hostname\n", program_name);
      exit(1);
    }

  time_hostname = argv[optind];

  /* Look up port number for time service. */
  service_info = getservbyname(TIME_SERVICE, "udp");
  if (service_info == NULL)
    {
      fprintf(stderr, "%s: " TIME_SERVICE "/udp: unknown service\n",
	      program_name);
      exit(1);
    }

  /* Resolve hostname (try MAX_TRIES times). We do this in case the
   * nameserver is just starting up (as in, everyone is just rebooting
   * from a long power failure, and we are requesting name resolution
   * before the server is ready).
   */
  for (tries = 0; tries < MAX_TRIES; tries++)
    {
      host_info = gethostbyname(time_hostname);
      if (host_info != NULL || (host_info == NULL && h_errno != TRY_AGAIN))
	break;
    }

  if (host_info == NULL)
    {
      fprintf(stderr, "%s: host %s unknown\n", program_name, time_hostname);
      exit(1);
    }

  if (host_info->h_addrtype != AF_INET)
    {
      fprintf(stderr, "%s: can't handle address type %d\n",
	      program_name, host_info->h_addrtype);
      exit(1);
    }

  /* Grab a socket. */
  time_socket = socket(AF_INET, SOCK_DGRAM, 0);
  if (time_socket < 0)
    myperror("socket", errno);

  /* Set up destination address. */
  time_address.sin_family = AF_INET;
  memcpy(&time_address.sin_addr, host_info->h_addr,
	 sizeof(time_address.sin_addr));
  time_address.sin_port = service_info->s_port;

  /* "Connect" the UDP socket to the time server address. */
  if (connect(time_socket, (struct sockaddr *)&time_address,
	      sizeof(time_address)) < 0)
    myperror("connect", errno);

  /* Initialize info for select. */
  FD_ZERO(&read_set);

  /* Attempt to acquire the time MAX_TRIES times at TIMEOUT interval. */
  for (tries = 0; tries < MAX_TRIES; tries++)
    {
      /* Just send an empty packet. */
      send(time_socket, buffer, 0, 0);

      /* Wait a little while for a reply. */
      FD_SET(time_socket, &read_set);
      timeout.tv_sec = TIMEOUT;
      timeout.tv_usec = 0;
      result = select(time_socket + 1, &read_set, NULL, NULL, &timeout);

      if (result == 0)		/* timed out, resend... */
	continue;

      if (result < 0)		/* error from select */
	myperror("select", errno);

      /* There must be data available. */
      result = recv(time_socket, buffer, sizeof(buffer), 0);
      if (result < 0)
	myperror("recv", errno);

      /* We should have received exactly four bytes in the packet. */
      if (result == 4)
	break;

      fprintf(stderr, "%s: received %d bytes, expected 4\n", program_name,
	      result);
      exit(1);
    }

  if (tries == MAX_TRIES)
    {
      fprintf(stderr, "%s: Failed to get time from %s.\n", program_name,
	      time_hostname);
      exit(1);
    }

  /* Convert RFC868 time to Unix time, and print it. */
  now = ((buffer[0] << 24) | (buffer[1] << 16) | (buffer[2] << 8) | buffer[3])
    - UNIX_OFFSET_TO_1900;
  fprintf(stdout, "%s", ctime(&now));

  /* Set the time if requested. */
  if (setflag)
    {
      gettimeofday(&current_time, &current_timezone);
      current_time.tv_sec = now;

      if (settimeofday(&current_time, &current_timezone) < 0)
	myperror("settimeofday", errno);
    }

  exit(0);
}
